docs/ide/api-designs/Workspace and Source Generated Documents.md

```diff
@@ -0,0 +1,73 @@
+
+This will run generators if they have not already ran, and then return a `SourceGeneratedDocument` for each generated document.
+
+The implementation of this will run `GetCompilationAsync` if the documents have not been generated; if the Compilation had already been generated, we would have cached the generated document information and so this would be cheap. We will hold onto the list of documents strongly (the tree will be a recoverable tree), so even if the `Compilation` is GC'ed we won't have to recompute this part a second time.
+
+### New API: `SourceGeneratedDocument? GetSourceGeneratedDocumentAsync(DocumentId, CancellationToken)`
+
+Fetches a single document by ID; equivalent to calling the API and filtering down to the right document.
+
+### Existing API: `GetDocument(SyntaxTree)`
+
+No changes, it will potentially return a `SourceGeneratedDocument` now. Callers _may_ have to check whether it's a generated document before they try to offer a fix or refactoring on the document.
+
+This API is the hardest API to figure out what to do in this entire spec. If it doesn't return generated documents, many features would break, since it's very common to assume that if a syntax tree exists in a `Compilation`, that there must be a `Document` that matches it. However, some features might _not_ want to see a generated document returned here because then they're going to try to modify that, and that won't make sense either. An audit of the code fixes in Roslyn discovered that for many of the places that would need an update to _ignore_ a generated document would also need an update to deal with the null reference that would come back from `GetDocument`.
+
+The other odd thing here is whether this function needs to be asynchronous or not. The current belief is no, because the only way you can have gotten a SyntaxTree to pass to it from the same `Solution` is either from:
+
+1. The `Compilation` or a symbol that came from it, which means generators have ran.
+2. You inspected some `SourceGeneratedDocument` and got it's tree, which means that generator has already ran.
+
+The only surprise you might run into is taking a `SyntaxTree` from an earlier `Solution`, passing it to a newer `Solution`, and getting a document that no longer exists because in the later `Solution` is no longer generating this document. However, passing a `SyntaxTree` between snapshots like that is questionable in the first place. Any code that is working with multiple snapshots and knew that a Document between the snapshots had the same tree could have done something like this, but that code is potentially broken _anyways_ with source generators since now one tree can change due to another tree changing.
+
+## APIs on `Workspace`
+
+### Existing APIs: `OpenDocument(DocumentId)`/`CloseDocument(DocumentId)`
+
+This API today is used by any feature that wants to tell the host to open a file. This will accept the DocumentId of a generated document and work properly.
```

Comment:
right.  because this is more an editor concept right?  just that a user/something is opening/viewing/interacting with the file.

Is the file considered readonly?

Reply:
> right. because this is more an editor concept right? just that a user/something is opening/viewing/interacting with the file.

I'm not sure I follow the question. These APIs are just the ones that a user of the workpace can use to open/close documents which just tells the host to do it. Just the abstraction over our services.

> Is the file considered readonly?

Yes.